File: 36-valid-sudoku/valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        cols = collections.defaultdict(set)
        rows = collections.defaultdict(set)
        squares = collections.defaultdict(set)

        for r in range(9):
            for c in range(9):
                current_value=board[r][c]
                if current_value=='.':
                    continue
                if current_value in rows[r] or current_value in cols[c] or current_value in squares[(r//3,c//3)]: 
                    return False
                cols[c].add(current_value)
                rows[r].add(current_value)
                squares[(r // 3,c // 3)].add(current_value)
        return True
        

File: 36-valid-sudoku/valid-sudoku.ts
function isValidSudoku(board: string[][]): boolean {
    const cols = new Map<number, Set<string>>()
    const rows = new Map<number, Set<string>>()
    const squares = new Map<number, Set<string>>()

    for (let r = 0; r < 9; r++){
        for(let c = 0; c < 9; c++){
            const num = board[r][c]
            if(num === '.') continue

            const squareIndex = Math.floor(r / 3) * 3 + Math.floor(c / 3)

            if(!rows[r]) rows[r] = new Set()
            if(!cols[c]) cols[c] = new Set()
            if(!squares[squareIndex]) squares[squareIndex] = new Set()

            if (rows[r].has(num) || cols[c].has(num) || squares[squareIndex].has(num)){
                return false
            }
            rows[r].add(num)
            cols[c].add(num)
            squares[squareIndex].add(num)
        }
    }
    return true
};